### PR TITLE
GLPK : GNU Linear Programming Kit

### PR DESCRIPTION
**Pull request by jjhelmus**

### CODE REVIEW

**Comment by jjhelmus** [comment]
Adapted from the conda/conda-recipes version.  @aebrahim would you be interested in being included as a maintainer of this recipe on conda-forge.  More information about the project can be found at the [conda-forge website](https://conda-forge.github.io/)

**Comment by aebrahim** [comment]
Sure!

I had never heard of conda-forge but it seems awesome.

Can I add my own packages to conda-forge as well?

**Comment by jakirkham** [comment]
Nice addition! Would be interested in being added too.

**Comment by jakirkham** [comment]
> Can I add my own packages to conda-forge as well?

Absolutely! Please do, @aebrahim. The docs are still a bit sparse, but the webpage has a very nice description on [adding recipes]( https://conda-forge.github.io/#add_recipe ).

**Comment by jjhelmus** [comment]
Looking to why this is failing.  The CircleCI and Travis issue I can probably solve but I have do idea on the AppVeyor.  Apparent the package is in conflict with itself...

**Comment by jjhelmus** [comment]
Can someone restart the AppVeyor build that failed?  I don't think it is real, merely a temporary HTTP error.

**Comment by jakirkham** [comment]
Looks green now. :)

File: recipes/glpk/meta.yaml [hunk 1]
```diff
@@ -11,10 +11,6 @@ source:
 
 build:
   number: 0
-  features:
-    - vc9      [win and py27]
```
Comment (aebrahim) [inline recipes/glpk/meta.yaml]: Without the vc features, how will you link python packages to glpk?

**Comment by jjhelmus** [comment]
@aebrahim Had a really good question that seems to have been suppressed. Is it alright to not have the `vc9` and similar features for the windows builds?  When I had them in the recipe, the tests failed as the package [conflicted with itself](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.189/job/qkfq241bjmr64dda).  @ocefpaf's Proj.4 recipe, #140, was using a `-python #[win]` build requirement to pull is the VS version which I think is sufficient.  From what I can tell usings features used to be required but no longer seems to be.

Also, I should mention that `make check` should be failing on OS X as libgmp.10.dylib is not in the correct location for the dynamic loaded to find it when running `./glpsol --version`. What I believe is happening is the homebrew version of libgmp is being picked up and loaded (same as what I think is happening in PR #132).  On my Mac the build does fail in `make check` as I do not have homebrew installed.  I do not think the LD_LIBRARY_PATH trick used in the run_test.sh file will work here as `make check` seems to ignore this variable.

The final libraries and glpsol executable work fine since conda adjusts the search paths with `install_name_tool` at the end of the build.  I'm just not sure how to trigger something similar during the build.

File: recipes/glpk/meta.yaml (hunk 1 above)
Comment (jakirkham) [inline recipes/glpk/meta.yaml]: @msarahan, do you have any thoughts on this?

**Comment by jakirkham** [comment]
> From what I can tell usings features used to be required but no longer seems to be.

I believe that is true. @msarahan, could you confirm?

**Comment by jakirkham** [comment]
> What I believe is happening is the homebrew version of libgmp is being picked up and loaded (same as what I think is happening in PR #132).

Ack! Good point. Maybe we should be cleaning everything (or nearly everything) out of brew.

File: recipes/glpk/meta.yaml (hunk 1 above)
Comment (aebrahim) [inline recipes/glpk/meta.yaml]: One way to confirm is to have a python build for the test, like in bioconda/bioconda-recipes#799

Comment (jjhelmus) [inline recipes/glpk/meta.yaml]: Yes, that was in your conda-recipes recipe but I remove when it was gaving me issues on my system.  I'll try to add it back it.

**Comment by jjhelmus** [comment]
Well that is unfortunate.  Apparently `conda build` will install a different version of python for the tests than the build...

File: recipes/glpk/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  fn: glpk-{{ version }}.tar.gz
+  url: http://ftp.gnu.org/gnu/glpk/glpk-{{ version }}.tar.gz
+  md5: c84ce7b8286ab91f2e871cd82050e2fe
+
+build:
+  number: 0
+
+requirements:
+  build:
+    - python   [win]
```
Comment (jakirkham) [inline recipes/glpk/meta.yaml]: Kind of hacky, but could try adding `x.x`.

File: recipes/glpk/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    - gmp      [not win]
+
+  run:
+    - gmp      [not win]
+
+test:
+  requires:
+    - python
```
Comment (jakirkham) [inline recipes/glpk/meta.yaml]: Then try adding `x.x` here. Not sure if it will work though. :/

**Comment by aebrahim** [comment]
Oh that's why I kept having issues! Do you know if continuum will fix this
bug? It's been so annoying having tests fail for seemingly no reason.

File: recipes/glpk/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  files:
+    - test.c
+    - setup.py
+    - spam.c
+
+  commands:
+    - glpsol --help
```
Comment (aebrahim) [inline recipes/glpk/meta.yaml]: why not ```glpsol --mps plan.mps``` (and include ```plan.mps``` in ```files```)

**Comment by jjhelmus** [comment]
Alright I figured out the features issue on Windows that was causing the package to conflict with itself during the test.  The `vc9`, `vc10`, and `vc14` features are only available after Python has been installed into an environment, so Python needs to be installed in the test environment or else there is an unsatisfiable dependency.  So although Python typically is not a dependency on Windows for GLPK it might as well be because the `vc` feature is not available without it.

File: recipes/glpk/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  fn: glpk-{{ version }}.tar.gz
+  url: http://ftp.gnu.org/gnu/glpk/glpk-{{ version }}.tar.gz
+  md5: c84ce7b8286ab91f2e871cd82050e2fe
+
+build:
+  number: 0
+
+requirements:
+  build:
+    - gmp      [not win]
```
Comment (jakirkham) [inline recipes/glpk/meta.yaml]: Shouldn't `python` still be a build dependency?

**Comment by jakirkham** [comment]
@msarahan, do you have any thoughts on how to better set the `vc` feature dependencies here? Would it be possible to do this *without* Python somehow?

**Comment by jjhelmus** [comment]
@aebrahim At this point the recipe is more or less the same as your conda-recipes version, just took a while to get there.

I think this is good go, if anyone has any final suggestions please give them.

**Comment by aebrahim** [comment]
:+1:

**Comment by jakirkham** [comment]
Does GLPK provide any options to support different sets of SSE/AVX instructions?

**Comment by jakirkham** [comment]
Once we have settled what to do on the above point, I am +1.

**Comment by jakirkham** [comment]
Here is some info I found for Linux. ( https://en.wikibooks.org/wiki/GLPK/Linux_OS#Advanced_configuration_options )

**Comment by aebrahim** [comment]
From the documentation I have read, it doesn't look like there are flags
for anything like that.

**Comment by aebrahim** [comment]
AFAIK, ```-march=native```, ```-march=sse2``` (etc.) are gcc flags which can be used for any program. I would think that we should use the same defaults as every other conda-forge package.

**Comment by jakirkham** [comment]
> AFAIK, `-march=native`, `-march=sse2` (etc.) are gcc flags which can be used for any program. 

True, they are.

> I would think that we should use the same defaults as every other conda-forge package.

I think you are right. Maybe we should consider some way to opt-in to more performant versions of GLPK, but that can certainly be done later.

@ocefpaf, would you mind giving this a once over and if it looks ok merge it? This will be a very nice library to have in conda-forge.

**Comment by aebrahim** [comment]
@jakirkham here's an idea I just had. If you want a super-performant sse4 version of glpk, wouldn't you want that for everything else too?

What about a second "conda-hyperforge" channel which only builds for 64 bit linux and by default sets ```$CLFAGS``` and ```$CPPFLAGS``` to ```-mavx2 -O3``` (etc.)? That way people with Haswell or later processors who need this can have all their packages built with those flags. (the downside being that this is quite a bit of extra work, and the slight performance tradeoff may not be worth it).

**Comment by msarahan** [comment]
@jjhelmus @aebrahim here's a recap of the feature stuff.  Aside from the content at https://github.com/conda/conda/wiki/VC-features, here's a bit more.

The key concept to features is that they are a way to limit any single process to only ever using one version of the MSVC runtime.  The process that should be limited in this way is the one that should have the track_feature.

Otherwise, for libraries, they'll generally have the build: features: section that the wiki page provides.  The fact that you also need Python in the build requirements is really pretty much a hack.  You need a way of specifying which compiler to use.  Since we (and the general Python community) tie compiler version to Python version strongly, it makes some (but not complete) sense to make the Python package be the thing that determines which compiler is used.  @ukoethe had a very good proposal of a different way to do it at https://github.com/conda/conda/issues/1959 - hopefully we'll work towards that.

Requiring Python as a runtime requirement will add py?? to the build string - otherwise, you only get the vc?? in the build string.

AFAIK, there is currently no way to have python behave correctly when it is a test requirement but not a runtime requirement.  The reason is that it is the runtime requirement that pins the python version.  This is separate from the vc feature, because any package built for a particular version of python would use the correct feature, by virtue of Python tracking the feature.

Where this really sucks is where you might have multiple programs each capable of tracking a feature of the same kind (two different vc features).  If they were both installed with different track features, then you'd have lots of unresolvable dependencies caused by ambiguous resolution.  You can, however, for standalone EXEs, manually specify different versions of the vs runtime (NOT features!)

For example, the nodejs build scripts automatically find VS 2015 on the system if it is installed.  Rather than limit nodejs to installation only with Python 3.5 environments, you can instead manually specify the vs2015_runtime package.  https://github.com/conda-forge/staged-recipes/pull/136/files  This does not conflict with any other vs20??_runtime package, and because nodejs is its own standalone exe, this works OK.  If nodejs depended on shared libraries that anything else might use, you could very easily run afoul of the no-mixed-runtime rule.  AFAICT, the only solution to that is to create one conda environment per process with shared libraries, or make the compiler version independent of python version (this is what @ukoethe's proposal does.

**Comment by msarahan** [comment]
@aebrahim I think it would be better to build multiple features.  The interesting thing about these features is that there can and probably should be fallbacks to less-optimized versions when more-optimized version are not available, rather than exclusive membership.  Default would be most compatible, but then have others for more optimizations.  The Numba team at Continuum is very interested in this.  I'll try to get them involved.

**Comment by jakirkham** [comment]
I wrote a response to your last comment @msarahan, but moved it to a [comment]( https://github.com/conda-forge/conda-forge.github.io/issues/49#issuecomment-198761613 ) in this issue ( https://github.com/conda-forge/conda-forge.github.io/issues/49 ). In general, in agreement with what you said above, but a few points to consider about infrastructure configuration, BLASes, packages that depend on them, potential licensing issues, etc.

**Comment by jakirkham** [comment]
It seems like we are all happy with the approach being taken here and CI clearly demonstrates the build is working. Thanks everyone for your hardwork and feedback. Merging...

**Comment by jakirkham** [comment]
The build issues with Python here might be related to this issue ( https://github.com/conda/conda-build/issues/835 ). Sorry had forgotten about it at the time. At least now it can be tracked.

**Pull request merged**